File: backend/app/utils/sentry_helpers.py
import sentry_sdk
from typing import Dict, Any, Optional, Callable
from functools import wraps
from fastapi import Request


def set_user_context(user_data: Dict[str, Any]) -> None:
    """
    Safely sets user context in Sentry, filtering out sensitive fields like passwords and tokens.

    Args:
        user_data (Dict[str, Any]): A dictionary containing user information.
    """
    try:
        # Filter out sensitive fields
        sensitive_fields = {'password', 'token', 'api_key', 'secret'}
        filtered_user_data = {k: v for k, v in user_data.items() if k not in sensitive_fields}
        sentry_sdk.set_user(filtered_user_data)
    except Exception:
        # Sentry not initialized or other error, silently ignore
        pass


def add_request_context(request: Request) -> None:
    """
    Extracts and sets request context in Sentry (method, URL, headers, query params) while filtering sensitive data.

    Args:
        request (Request): The FastAPI request object.
    """
    try:
        # Filter sensitive headers
        sensitive_headers = {'authorization', 'cookie', 'x-api-key'}
        filtered_headers = {k: v for k, v in request.headers.items() if k.lower() not in sensitive_headers}
        
        sentry_sdk.set_context("request", {
            "method": request.method,
            "url": str(request.url),
            "headers": filtered_headers,
            "query_params": dict(request.query_params)
        })
    except Exception:
        # Sentry not initialized or other error, silently ignore
        pass


def capture_message_with_context(message: str, level: str, **context: Any) -> None:
    """
    Captures a message with custom context in Sentry.

    Args:
        message (str): The message to capture.
        level (str): The log level (e.g., 'info', 'warning', 'error').
        **context: Additional context key-value pairs.
    """
    try:
        sentry_sdk.capture_message(message, level=level, extra=context)
    except Exception:
        # Sentry not initialized or other error, silently ignore
        pass


def start_transaction(name: str, op: str) -> Optional[Any]:
    """
    Starts a Sentry transaction and returns it for manual instrumentation.

    Args:
        name (str): The name of the transaction.
        op (str): The operation type.

    Returns:
        Optional[Any]: The Sentry transaction object, or None if Sentry is not initialized.
    """
    try:
        return sentry_sdk.start_transaction(name=name, op=op)
    except Exception:
        # Sentry not initialized or other error, return None
        return None


def monitor_performance(operation: str) -> Callable:
    """
    Decorator that wraps functions with Sentry spans for performance monitoring.

    Args:
        operation (str): The operation type for the span.

    Returns:
        Callable: The decorated function.
    """
    def decorator(func: Callable) -> Callable:
        @wraps(func)
        def wrapper(*args: Any, **kwargs: Any) -> Any:
            try:
                with sentry_sdk.start_span(op=operation, description=func.__name__):
                    return func(*args, **kwargs)
            except Exception:
                # If Sentry span fails, still execute the function
                return func(*args, **kwargs)
        return wrapper
    return decorator


def configure_scope_for_request(request: Request, user: Optional[Dict[str, Any]] = None) -> None:
    """
    Sets up Sentry scope with request and user context.

    Args:
        request (Request): The FastAPI request object.
        user (Optional[Dict[str, Any]]): Optional user data dictionary.
    """
    try:
        with sentry_sdk.configure_scope() as scope:
            # Set request context
            sensitive_headers = {'authorization', 'cookie', 'x-api-key'}
            filtered_headers = {k: v for k, v in request.headers.items() if k.lower() not in sensitive_headers}
            
            scope.set_context("request", {
                "method": request.method,
                "url": str(request.url),
                "headers": filtered_headers,
                "query_params": dict(request.query_params)
            })
            
            # Set user context if provided
            if user:
                set_user_context(user)
    except Exception:
        # Sentry not initialized or other error, silently ignore
        pass

File: src/lib/sentry-helpers.ts
import * as Sentry from '@sentry/react';

// Define the User interface for type safety
interface User {
  id: string;
  email: string;
  username?: string;
  role?: string;
  // Add other non-sensitive fields as needed
}

/**
 * Safely sets the user context in Sentry, filtering out any sensitive fields.
 * If Sentry is not initialized, this function does nothing.
 * @param user - The user object or null to clear context
 */
export function setUserContext(user: User | null): void {
  try {
    if (user) {
      Sentry.setUser({
        id: user.id,
        email: user.email,
        username: user.username,
        // Add other non-sensitive fields, e.g., role: user.role
      });
    } else {
      Sentry.setUser(null);
    }
  } catch (error) {
    // Sentry not initialized or other error, silently ignore
  }
}

/**
 * Captures an API error with relevant context.
 * If Sentry is not initialized, this function does nothing.
 * @param error - The error object
 * @param endpoint - The API endpoint
 * @param method - The HTTP method
 * @param statusCode - Optional HTTP status code
 */
export function captureApiError(error: Error, endpoint: string, method: string, statusCode?: number): void {
  try {
    Sentry.captureException(error, {
      contexts: {
        api: {
          endpoint,
          method,
          statusCode,
        },
      },
    });
  } catch (err) {
    // Sentry not initialized or other error, silently ignore
  }
}

/**
 * Adds a breadcrumb for route navigation.
 * If Sentry is not initialized, this function does nothing.
 * @param from - The source route
 * @param to - The destination route
 */
export function addNavigationBreadcrumb(from: string, to: string): void {
  try {
    Sentry.addBreadcrumb({
      message: `Navigation from ${from} to ${to}`,
      category: 'navigation',
      level: 'info',
    });
  } catch (error) {
    // Sentry not initialized or other error, silently ignore
  }
}

/**
 * Measures the performance of a callback function using Sentry spans.
 * If Sentry is not initialized, the callback is executed without measurement.
 * @param name - The name of the performance measurement
 * @param callback - The function to measure
 */
export function measurePerformance(name: string, callback: () => void): void {
  try {
    Sentry.startSpan({ name, op: 'function' }, () => {
      callback();
    });
  } catch (error) {
    // Sentry not initialized or other error, execute callback without measurement
    callback();
  }
}

/**
 * Captures a form submission error with sanitized form data.
 * Sensitive fields like 'password', 'token', 'apiKey' are removed from formData.
 * If Sentry is not initialized, this function does nothing.
 * @param formName - The name of the form
 * @param error - The error object
 * @param formData - Optional form data to include (will be sanitized)
 */
export function captureFormError(formName: string, error: Error, formData?: Record<string, any>): void {
  try {
    const sensitiveKeys = ['password', 'token', 'apikey', 'secret', 'key'];
    const sanitizedFormData = formData
      ? Object.keys(formData).reduce((acc, key) => {
          if (!sensitiveKeys.includes(key.toLowerCase())) {
            acc[key] = formData[key];
          }
          return acc;
        }, {} as Record<string, any>)
      : undefined;

    Sentry.captureException(error, {
      contexts: {
        form: {
          formName,
          formData: sanitizedFormData,
        },
      },
    });
  } catch (err) {
    // Sentry not initialized or other error, silently ignore
  }
}

/**
 * Starts a manual Sentry transaction.
 * If Sentry is not initialized, this function returns null.
 * @param name - The name of the transaction
 * @param op - The operation type
 * @returns The transaction object or null if Sentry is not initialized
 */
export function startTransaction(name: string, op: string): Sentry.Transaction | null {
  try {
    return Sentry.startTransaction({ name, op });
  } catch (error) {
    // Sentry not initialized or other error, return null
    return null;
  }
}